compiler/rustc_typeck/src/check/place_op.rs

```diff
@@ -82,6 +86,53 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             expr, base_expr, adjusted_ty, index_ty
         );
 
+        let negative_index = || {
```

Comment:
Could you add a sibling function for this logic rather than a closure? It's nice if we can keep the diagnostic logic as off-to-the-side as possible and out of the main functions so as not to impede understanding of the function's purpose.